### PR TITLE
Code Coverage Unit Tests for comm.h

### PR DESCRIPTION
**Pull request by ishkool**
## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", Added new unit tests to cover test cases that were not covered by the functional tests for src/include/comm.h file.

**What were the changes?**  
Addition of new unit tests

**Why were the changes made?**  
The motivation behind this task was to improve the overall code quality.
**Added 1 test for uncovered function, the functional code coverage is 100% (8/8) from previous 87.5%(7/8)**

**How was the outcome achieved?**  
_Technical details behind the work. Explain any publicly-available hardware peculiarities._

**Additional Documentation:**  
_What else should the reviewer know?_

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Comment by venksubramd** [comment]
The test file currently has one test that checks the behavior of the ncclTaskCollSorterEmpty function. The test may be simplified. Instead of creating the instance of ncclTaskCollSorter using new and deleting it, we may create it on the stack and pass a pointer to it to the ncclTaskCollSorterEmpty function.

The ncclTaskCollSorterEmpty function does not handle the case of the pointer to ncclTaskCollSorter being empty. Should this function be returning an error rather than crashing in such case?

Do we need tests for
ncclTaskCollSorterInsert
ncclTaskCollSorterDequeueAll
ncclCommPollCallbacks
ncclCommPollEventCallbacks
ncclCommIntraBarrierIn
ncclCommIntraBarrierOut
ncclUserRedOpMangle
?

**Comment by ishkool** [comment]
> The test file currently has one test that checks the behavior of the ncclTaskCollSorterEmpty function. The test may be simplified. Instead of creating the instance of ncclTaskCollSorter using new and deleting it, we may create it on the stack and pass a pointer to it to the ncclTaskCollSorterEmpty function.
> 
> The ncclTaskCollSorterEmpty function does not handle the case of the pointer to ncclTaskCollSorter being empty. Should this function be returning an error rather than crashing in such case?
> 
> Do we need tests for ncclTaskCollSorterInsert ncclTaskCollSorterDequeueAll ncclCommPollCallbacks ncclCommPollEventCallbacks ncclCommIntraBarrierIn ncclCommIntraBarrierOut ncclUserRedOpMangle ?

1) Some structs are huge and cause a stack overflow if directly declared on the stack, I have witnessed it several times hence I have adopted the heap route. Since compilation takes a lot of time hence I'm following this for everything while making sure I free up the memory afterwards.
2) No error checking is required since ncclTaskCollSorterEmpty just checks if the pointer is a nullptr or not and returns a bool value accordingly.
3) The other mentions functions are already being called/covered and hence no tests are required for those. This is mostly for code coverage activity.

Thank you.

**Comment by atulkulk** [comment]
> > The test file currently has one test that checks the behavior of the ncclTaskCollSorterEmpty function. The test may be simplified. Instead of creating the instance of ncclTaskCollSorter using new and deleting it, we may create it on the stack and pass a pointer to it to the ncclTaskCollSorterEmpty function.
> > The ncclTaskCollSorterEmpty function does not handle the case of the pointer to ncclTaskCollSorter being empty. Should this function be returning an error rather than crashing in such case?
> > Do we need tests for ncclTaskCollSorterInsert ncclTaskCollSorterDequeueAll ncclCommPollCallbacks ncclCommPollEventCallbacks ncclCommIntraBarrierIn ncclCommIntraBarrierOut ncclUserRedOpMangle ?
> 
> 1. Some structs are huge and cause a stack overflow if directly declared on the stack, I have witnessed it several times hence I have adopted the heap route. Since compilation takes a lot of time hence I'm following this for everything while making sure I free up the memory afterwards.
> 2. No error checking is required since ncclTaskCollSorterEmpty just checks if the pointer is a nullptr or not and returns a bool value accordingly.
> 3. The other mentions functions are already being called/covered and hence no tests are required for those. This is mostly for code coverage activity.

@ishkool Let me add some clarification on point 3, the goal here is not just code coverage activity but to have meaningful tests which will help us find any regressions or bugs early in the development cycle. Code coverage metrics does help us in identifying which code paths were taken and which weren't.

> Thank you.

File: test/CommTest.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+{
+  TEST(CommTest, Sorter)
+  {
+	TestBed testBed;
+	INFO("[CommTest] starting the test with %d GPUs\n", testBed.numDevicesAvailable);
+	// Configuration
+	ncclTaskCollSorter* me_ptr = new ncclTaskCollSorter;
+	me_ptr->head = nullptr;
+
+	assert(true == ncclTaskCollSorterEmpty(me_ptr));
```
Comment (atulkulk) [inline test/CommTest.cpp]: Use GTest API for equality checks.

**Comment by venksubramd** [comment]
@ishkool

> 1. Some structs are huge and cause a stack overflow if directly declared on the stack, I have witnessed it several times hence I have adopted the heap route. Since compilation takes a lot of time hence I'm following this for everything while making sure I free up the memory afterwards.

Thanks you for providing the reason. It may be better to treat this as an exception than a norm. We can keep the tests less noisy and simple by keeping things on the stack and only use the heap when really necessary. Clarity and cleanliness over convenience is something that may be useful to consider. If we are running into heap issues most of the time, then the option to use new/malloc and delete may be reasonable.

**Comment by ishkool** [comment]
> > > The test file currently has one test that checks the behavior of the ncclTaskCollSorterEmpty function. The test may be simplified. Instead of creating the instance of ncclTaskCollSorter using new and deleting it, we may create it on the stack and pass a pointer to it to the ncclTaskCollSorterEmpty function.
> > > The ncclTaskCollSorterEmpty function does not handle the case of the pointer to ncclTaskCollSorter being empty. Should this function be returning an error rather than crashing in such case?
> > > Do we need tests for ncclTaskCollSorterInsert ncclTaskCollSorterDequeueAll ncclCommPollCallbacks ncclCommPollEventCallbacks ncclCommIntraBarrierIn ncclCommIntraBarrierOut ncclUserRedOpMangle ?
> > 
> > 
> > 
> > 1. Some structs are huge and cause a stack overflow if directly declared on the stack, I have witnessed it several times hence I have adopted the heap route. Since compilation takes a lot of time hence I'm following this for everything while making sure I free up the memory afterwards.
> > 2. No error checking is required since ncclTaskCollSorterEmpty just checks if the pointer is a nullptr or not and returns a bool value accordingly.
> > 3. The other mentions functions are already being called/covered and hence no tests are required for those. This is mostly for code coverage activity.
> 
> @ishkool Let me add some clarification on point 3, the goal here is not just code coverage activity but to have meaningful tests which will help us find any regressions or bugs early in the development cycle. Code coverage metrics does help us in identifying which code paths were taken and which weren't.

@atulkulk Agreed but these tests are already being covered so no point in writing redundant tests. 

> 
> > Thank you.

**Review by rahulvaidya20 — APPROVED** [review]
Looks okay for now. We can also add another test for non-nullptr case or other edge cases later to check if the function exits gracefully.

**Comment by venksubramd** [comment]
Looks good for now. As @rahulvaidya20 has suggested we should consider those additional tests later. OK to close.

**Review by PJAvinash — APPROVED** [review]

**Review by Nikhil-Nunna — APPROVED** [review]
